ts/packages/ts/src/utils.ts

```diff
@@ -1,24 +1,56 @@
 import _ from 'lodash';
 import Mustache from 'mustache';
+import Ajv, { SchemaObject } from 'ajv';
 
-export const ERR = (
-  message: string,
-  { location = [], suggestion = '' }: { location?: string[]; suggestion?: string } = {},
-) => {
-  return `${message}${!_.isEmpty(location) ? ` at ${location.join('.')}` : ''}${suggestion ? `, ${suggestion}` : ''}`;
-};
+export class ConfigError extends Error {
+  constructor(
+    public reason: string,
+    public hint: string = '', // code?: string,
+    public scope: [string, string][] = [],
+  ) {
+    super();
+    this.setMessage();
+  }
+
+  protected setMessage() {
+    this.message = this.reason;
+    if (!_.isEmpty(this.scope)) {
+      this.message = `${this.message} at ${this.scope.map(([domain, name]) => `${domain}:${name}`).join('.')}`;
+    }
+    if (this.hint) {
+      this.message = `${this.message}, ${this.hint}`;
+    }
+  }
+
+  setReason(reason: string) {
+    this.reason = reason;
+    this.setMessage();
+  }
+
+  setHint(hint: string) {
+    this.hint = hint;
+    this.setMessage();
+  }
+
+  appendScope(scope: [string, string][]) {
+    this.scope = [...scope, ...this.scope];
+    this.setMessage();
+  }
```

Comment:
Really not a fan of having mutable error classes like this. For one it makes it more difficult to trace the of the error to the lowest level it was raised, secondly it doesn't play nicely at all with error reporting tools like [Sentry](https://sentry.io/welcome/). Generally better than mutating an error is to create a new error from the existing one.

Reply:
ABSOLUTELY!